### ISSUE TITLE
prte cleans the whole job when all job-local procs terminate at a given prted

### ISSUE THREAD

**Issue by abouteiller** [openpmix/prrte#1698]
As soon as **all** procs on a particular PRTED have completed, the entire job is dismantled (and the KILL command is sent to the remainder of the application procs for the job at other PRTEDs).

This is problematic for resilience (e.g., for RECOVERABLE jobs), but is also probably incorrect for the (failure-free) case of applications that do not employ synchronized finalize routines (e.g., MPI_Finalize in fenceless mode, if procs on one node all call finalize and exit, this will cause the other procs to be killed immediately). 

If at least one application rank is still active at any given prted, the behavior is correct (cleanup is delayed until the remaining procs terminate on their own).

## Background information

 a75f933a (ompi/main, main)
 22dabb8e471fb110ca75a5fd3d5f12d8d6a984a3 3rd-party/openpmix (v1.1.3-3791-g22dabb8e)
+3bac3acda2d751e93eec6553b25464749d1b5fd5 3rd-party/prrte (psrvr-v2.0.0rc1-4606-g3bac3acd)
 237ceff1a8ed996d855d69f372be9aaea44919ea config/oac (remotes/origin/HEAD)

### Please describe the system on which you are running

* Operating system/version: CentOS/7
* Computer hardware: x86_64
* Network type: TCP

-----------------------------

## Details of the problem

1. Using the ULFM testers `git@github.com:ICLDisco/ulfm-testing.git` I have the `tutorial/02.err_handler` program deployed on 2 nodes, 2 ranks/node. 
2. The tester kills rank 2 and rank 3 (both located on daemon at node 'b05').
3. After discovering that all its local procs are gone, b05 prted sends the command `[b05.bezout:140057] [prterun-saturn-65366@0,2] state:prted: SENDING JOB LOCAL TERMINATION UPDATE FOR JOB prterun-saturn-65366@1` 
4. This in turn causes `mpiexec` to trigger a job cleanup. 
```
[saturn.icl.utk.edu:12018] [prterun-saturn-12018@0,0] [1678483834.010748] ACTIVATE JOB prterun-saturn-12018@1 STATE NOTIFIED AT ../../../../../../../master/3rd-party/prrte/src/mca/state/dvm/state_dvm.c:1071
[saturn.icl.utk.edu:12018] [prterun-saturn-12018@0,0] [1678483834.010762] ACTIVATING JOB prterun-saturn-12018@1 STATE NOTIFIED PRI 4
[saturn.icl.utk.edu:12018] [prterun-saturn-12018@0,0] odls:kill_local_proc working on WILDCARD
[b00.bezout:98129] [prterun-saturn-12018@0,1] odls:kill_local_proc working on WILDCARD
[b05.bezout:140449] [prterun-saturn-12018@0,2] odls:kill_local_proc working on WILDCARD
```

The same scenario with 3 ranks/node (and still 2 faults) operates correctly (that is, the job cleanup triggers only after all surviving procs call MPI_Finalize, because there is at least one live proc on each prted).

A full piece of the relevant part of the log:
```
[b05.bezout:140449] [prterun-saturn-12018@0,2] state:prted:track_procs called for proc [prterun-saturn-12018@1,2] state NORMALLY TERMINATED
Rank 02 / 04: bye bye!
[saturn.icl.utk.edu:12018] [prterun-saturn-12018@0,0] [1678483834.006743] ACTIVATING PROC [prterun-saturn-12018@1,2] STATE ABORTED BY SIGNAL PRI 0
Rank 03 / 04: bye bye!
[saturn.icl.utk.edu:12018] [prterun-saturn-12018@0,0] errmgr:dvm: for proc [prterun-saturn-12018@1,2] state ABORTED BY SIGNAL
[saturn.icl.utk.edu:12018] [prterun-saturn-12018@0,0] [1678483834.006810] ACTIVATE PROC [prterun-saturn-12018@1,2] STATE WAITPID FIRED AT ../../../../../../../master/3rd-party/prrte/src/mca/errmgr/dvm/errmgr_dvm.c:377
[saturn.icl.utk.edu:12018] [prterun-saturn-12018@0,0] [1678483834.006825] ACTIVATING PROC [prterun-saturn-12018@1,2] STATE WAITPID FIRED PRI 4
[saturn.icl.utk.edu:12018] [prterun-saturn-12018@0,0] [1678483834.006842] ACTIVATE PROC [prterun-saturn-12018@1,2] STATE IOF COMPLETE AT ../../../../../../../master/3rd-party/prrte/src/mca/errmgr/dvm/errmgr_dvm.c:383
[saturn.icl.utk.edu:12018] [prterun-saturn-12018@0,0] [1678483834.006855] ACTIVATING PROC [prterun-saturn-12018@1,2] STATE IOF COMPLETE PRI 4
[saturn.icl.utk.edu:12018] [prterun-saturn-12018@0,0] errmgr:dvm: proc [prterun-saturn-12018@1,2] aborted by signal
[saturn.icl.utk.edu:12018] [prterun-saturn-12018@0,0] errmgr:dvm:sending notification PROC-ABORTED-BY-SIG affected proc [prterun-saturn-12018@1,2]
[saturn.icl.utk.edu:12018] [prterun-saturn-12018@0,0] state:base:track_procs called for proc [prterun-saturn-12018@1,2] state WAITPID FIRED
[saturn.icl.utk.edu:12018] [prterun-saturn-12018@0,0] state:base:track_procs called for proc [prterun-saturn-12018@1,2] state IOF COMPLETE
[saturn.icl.utk.edu:12018] [prterun-saturn-12018@0,0] [1678483834.007347] ACTIVATE PROC [prterun-saturn-12018@1,2] STATE NORMALLY TERMINATED AT ../../../../../../master/3rd-party/prrte/src/mca/state/base/state_base_fns.c:553
[saturn.icl.utk.edu:12018] [prterun-saturn-12018@0,0] [1678483834.007364] ACTIVATING PROC [prterun-saturn-12018@1,2] STATE NORMALLY TERMINATED PRI 4
[saturn.icl.utk.edu:12018] [prterun-saturn-12018@0,0] state:base:track_procs called for proc [prterun-saturn-12018@1,2] state NORMALLY TERMINATED
[b05.bezout:140449] [prterun-saturn-12018@0,2] [1678483834.008360] ACTIVATE PROC [prterun-saturn-12018@1,3] STATE IOF COMPLETE AT ../../../../../../../master/3rd-party/prrte/src/mca/iof/prted/iof_prted_read.c:197
[b05.bezout:140449] [prterun-saturn-12018@0,2] [1678483834.008374] ACTIVATING PROC [prterun-saturn-12018@1,3] STATE IOF COMPLETE PRI 4
[b05.bezout:140449] [prterun-saturn-12018@0,2] state:prted:track_procs called for proc [prterun-saturn-12018@1,3] state IOF COMPLETE
[b05.bezout:140449] [prterun-saturn-12018@0,2] odls:wait_local_proc child process [prterun-saturn-12018@1,3] pid 140496 terminated
[b05.bezout:140449] [prterun-saturn-12018@0,2] odls:waitpid_fired child process [prterun-saturn-12018@1,3] terminated with signal Killed
[b05.bezout:140449] [prterun-saturn-12018@0,2] [1678483834.008597] ACTIVATE PROC [prterun-saturn-12018@1,3] STATE ABORTED BY SIGNAL AT ../../../../../../master/3rd-party/prrte/src/mca/odls/base/odls_base_default_fns.c:1717
[b05.bezout:140449] [prterun-saturn-12018@0,2] [1678483834.008616] ACTIVATING PROC [prterun-saturn-12018@1,3] STATE ABORTED BY SIGNAL PRI 0
[b05.bezout:140449] [prterun-saturn-12018@0,2] errmgr:prted:proc_errors process [prterun-saturn-12018@1,3] error state ABORTED BY SIGNAL
[b05.bezout:140449] [prterun-saturn-12018@0,2] errmgr:prted got state ABORTED BY SIGNAL for proc [prterun-saturn-12018@1,3]
[saturn.icl.utk.edu:12018] [prterun-saturn-12018@0,0] [1678483834.009395] ACTIVATE PROC [prterun-saturn-12018@1,3] STATE ABORTED BY SIGNAL AT ../../../../../../master/3rd-party/prrte/src/mca/plm/base/plm_base_receive.c:426
[b05.bezout:140449] [prterun-saturn-12018@0,2] errmgr:prted reporting proc [prterun-saturn-12018@1,3] abprted to HNP (local procs = 2)
[b05.bezout:140449] [prterun-saturn-12018@0,2] [1678483834.008664] ACTIVATE PROC [prterun-saturn-12018@1,3] STATE NORMALLY TERMINATED AT ../../../../../../../master/3rd-party/prrte/src/mca/errmgr/prted/errmgr_prted.c:646
[b05.bezout:140449] [prterun-saturn-12018@0,2] [1678483834.008671] ACTIVATING PROC [prterun-saturn-12018@1,3] STATE NORMALLY TERMINATED PRI 4
[saturn.icl.utk.edu:12018] [prterun-saturn-12018@0,0] [1678483834.009440] ACTIVATING PROC [prterun-saturn-12018@1,3] STATE ABORTED BY SIGNAL PRI 0
[b05.bezout:140449] [prterun-saturn-12018@0,2] state:prted:track_procs called for proc [prterun-saturn-12018@1,3] state NORMALLY TERMINATED
[saturn.icl.utk.edu:12018] [prterun-saturn-12018@0,0] errmgr:dvm: for proc [prterun-saturn-12018@1,3] state ABORTED BY SIGNAL
[b05.bezout:140449] [prterun-saturn-12018@0,2] state:prted: SENDING JOB LOCAL TERMINATION UPDATE FOR JOB prterun-saturn-12018@1
[saturn.icl.utk.edu:12018] [prterun-saturn-12018@0,0] [1678483834.009483] ACTIVATE PROC [prterun-saturn-12018@1,3] STATE WAITPID FIRED AT ../../../../../../../master/3rd-party/prrte/src/mca/errmgr/dvm/errmgr_dvm.c:377
[saturn.icl.utk.edu:12018] [prterun-saturn-12018@0,0] [1678483834.009499] ACTIVATING PROC [prterun-saturn-12018@1,3] STATE WAITPID FIRED PRI 4
[saturn.icl.utk.edu:12018] [prterun-saturn-12018@0,0] [1678483834.009521] ACTIVATE PROC [prterun-saturn-12018@1,3] STATE IOF COMPLETE AT ../../../../../../../master/3rd-party/prrte/src/mca/errmgr/dvm/errmgr_dvm.c:383
[saturn.icl.utk.edu:12018] [prterun-saturn-12018@0,0] [1678483834.009539] ACTIVATING PROC [prterun-saturn-12018@1,3] STATE IOF COMPLETE PRI 4
[saturn.icl.utk.edu:12018] [prterun-saturn-12018@0,0] errmgr:dvm: proc [prterun-saturn-12018@1,3] aborted by signal
[saturn.icl.utk.edu:12018] [prterun-saturn-12018@0,0] errmgr:dvm:sending notification PROC-ABORTED-BY-SIG affected proc [prterun-saturn-12018@1,3]
[b05.bezout:140449] [prterun-saturn-12018@0,2] state:prted releasing procs from node b00
[b05.bezout:140449] [prterun-saturn-12018@0,2] state:prted releasing proc [prterun-saturn-12018@1,0] from node b00
[b05.bezout:140449] [prterun-saturn-12018@0,2] state:prted releasing proc [prterun-saturn-12018@1,1] from node b00
[b05.bezout:140449] [prterun-saturn-12018@0,2] state:prted releasing procs from node b05
[b05.bezout:140449] [prterun-saturn-12018@0,2] state:prted releasing proc [prterun-saturn-12018@1,2] from node b05
[b05.bezout:140449] [prterun-saturn-12018@0,2] state:prted releasing proc [prterun-saturn-12018@1,3] from node b05
[saturn.icl.utk.edu:12018] [prterun-saturn-12018@0,0] [1678483834.009680] ACTIVATE PROC [prterun-saturn-12018@1,2] STATE NORMALLY TERMINATED AT ../../../../../../master/3rd-party/prrte/src/mca/plm/base/plm_base_receive.c:426
[saturn.icl.utk.edu:12018] [prterun-saturn-12018@0,0] [1678483834.009699] ACTIVATING PROC [prterun-saturn-12018@1,2] STATE NORMALLY TERMINATED PRI 4
[saturn.icl.utk.edu:12018] [prterun-saturn-12018@0,0] [1678483834.009717] ACTIVATE PROC [prterun-saturn-12018@1,3] STATE NORMALLY TERMINATED AT ../../../../../../master/3rd-party/prrte/src/mca/plm/base/plm_base_receive.c:426
[saturn.icl.utk.edu:12018] [prterun-saturn-12018@0,0] [1678483834.009730] ACTIVATING PROC [prterun-saturn-12018@1,3] STATE NORMALLY TERMINATED PRI 4
[saturn.icl.utk.edu:12018] [prterun-saturn-12018@0,0] state:base:track_procs called for proc [prterun-saturn-12018@1,3] state WAITPID FIRED
[saturn.icl.utk.edu:12018] [prterun-saturn-12018@0,0] state:base:track_procs called for proc [prterun-saturn-12018@1,3] state IOF COMPLETE
[saturn.icl.utk.edu:12018] [prterun-saturn-12018@0,0] [1678483834.009902] ACTIVATE PROC [prterun-saturn-12018@1,3] STATE NORMALLY TERMINATED AT ../../../../../../master/3rd-party/prrte/src/mca/state/base/state_base_fns.c:553
[saturn.icl.utk.edu:12018] [prterun-saturn-12018@0,0] [1678483834.009917] ACTIVATING PROC [prterun-saturn-12018@1,3] STATE NORMALLY TERMINATED PRI 4
[saturn.icl.utk.edu:12018] [prterun-saturn-12018@0,0] state:base:track_procs called for proc [prterun-saturn-12018@1,2] state NORMALLY TERMINATED
[saturn.icl.utk.edu:12018] [prterun-saturn-12018@0,0] state:base:track_procs called for proc [prterun-saturn-12018@1,3] state NORMALLY TERMINATED
[saturn.icl.utk.edu:12018] [prterun-saturn-12018@0,0] state:base:track_procs called for proc [prterun-saturn-12018@1,3] state NORMALLY TERMINATED
[saturn.icl.utk.edu:12018] [prterun-saturn-12018@0,0] [1678483834.009975] ACTIVATE JOB prterun-saturn-12018@1 STATE NORMALLY TERMINATED AT ../../../../../../master/3rd-party/prrte/src/mca/state/base/state_base_fns.c:623
[saturn.icl.utk.edu:12018] [prterun-saturn-12018@0,0] [1678483834.009994] ACTIVATING JOB prterun-saturn-12018@1 STATE NORMALLY TERMINATED PRI 4
[saturn.icl.utk.edu:12018] [prterun-saturn-12018@0,0] state:dvm:check_job_complete on job prterun-saturn-12018@1
[saturn.icl.utk.edu:12018] [prterun-saturn-12018@0,0] state:dvm:check_job_completed state is terminated - activating notify
[saturn.icl.utk.edu:12018] [prterun-saturn-12018@0,0] [1678483834.010650] ACTIVATE JOB prterun-saturn-12018@1 STATE NOTIFY COMPLETED AT ../../../../../../../master/3rd-party/prrte/src/mca/state/dvm/state_dvm.c:671
[saturn.icl.utk.edu:12018] [prterun-saturn-12018@0,0] [1678483834.010667] ACTIVATING JOB prterun-saturn-12018@1 STATE NOTIFY COMPLETED PRI 4
[saturn.icl.utk.edu:12018] [prterun-saturn-12018@0,0] state:dvm:dvm_notify called
[saturn.icl.utk.edu:12018] [prterun-saturn-12018@0,0] state:dvm:dvm_notify notification requested
[saturn.icl.utk.edu:12018] [prterun-saturn-12018@0,0] state:dvm:dvm_notify notification sent
[saturn.icl.utk.edu:12018] [prterun-saturn-12018@0,0] [1678483834.010748] ACTIVATE JOB prterun-saturn-12018@1 STATE NOTIFIED AT ../../../../../../../master/3rd-party/prrte/src/mca/state/dvm/state_dvm.c:1071
[saturn.icl.utk.edu:12018] [prterun-saturn-12018@0,0] [1678483834.010762] ACTIVATING JOB prterun-saturn-12018@1 STATE NOTIFIED PRI 4
[saturn.icl.utk.edu:12018] [prterun-saturn-12018@0,0] odls:kill_local_proc working on WILDCARD
[b00.bezout:98129] [prterun-saturn-12018@0,1] odls:kill_local_proc working on WILDCARD
[b05.bezout:140449] [prterun-saturn-12018@0,2] odls:kill_local_proc working on WILDCARD

```

Launched with 
```
salloc -N 2 -Cbezout  ~/ompi/master.debug/bin/mpiexec --prefix ~/ompi/master.debug -N 2  --mca btl_tcp_if_include ib0 --mca btl tcp,self --with-ft ulfm --display bindings --display-comm  --mca ompi_mpi_ft_verbose 5  --leave-session-attached  --mca errmgr_base_verbose 100 --mca state_base_verbose  100  --mca odls_base_verbose 100 --mca plm_base_verbose 00 --mca oob_base_verbose 0 --mca opal_abort_delay -1    --mca pmix_server_get_verbose 00 --mca pml_base_verbose 2 --mca btl_base_warn_peer_error false  02.err_handler
```

**Comment by rhc54** [comment]
> As soon as all procs on a particular PRTED have completed, the entire job is dismantled (and the KILL command is sent to the remainder of the application procs for the job at other PRTEDs).

I'm not seeing that behavior, but I suspect one difference is that you are killing the proc by signal instead of it aborting itself. Just guessing. Still, the title here is misleading as it appears you are seeing something to do with a very specific use-case, and not job termination just because the procs on a given node complete.

**Comment by rhc54** [comment]
The problem here was created by your last commit. As I indicated on that PR, I had reservations about its correctness - my concern was regarding the accounting system and whether PRRTE would be counting terminations correctly. Turns out that concern was correct - the patch resulted in double-counting of proc termination, thereby causing mpirun to believe that all procs had completed once half of them had done so (didn't matter which procs - normally terminated or killed - so long as half completed).

https://github.com/openpmix/prrte/pull/1701 fixes that problem, but I don't consider it a complete solution. The lingering problem is that the daemon where the procs were killed sees the local job as "complete", and therefore it locally cleans up the entire job record. This includes deregistering the job from its PMIx server library and releasing all resources and record associated with other procs on other nodes of the job.

If the job is recoverable and procs are restarted on this node, then the loss of that data will have consequences and probably lead to incorrect information being passed to the MPI library. I suspect you don't want this cleanup to occur - we probably should hold off on doing it until the HNP tells us to cleanup the job.

So I think the complete fix requires that we "deregister client" on each proc as it terminates (normally or not), and hold off on "deregister nspace" and cleaning up resources until the HNP tells us "cleanup this job". I'll leave that to you. If your goal is just to let the job shrink and not restart procs, then the fix in #1701 should be adequate.

**Comment by rhc54** [comment]
I believe this is "good enough" for now - please reopen if necessary.

**Comment by abouteiller** [comment]
I should have informed you of intermediate progress earlier. 

I could not verify If that does fix the problem for me (I believe it will though) because that has brought back another issue where pmix_fence does not complete during MPI_Finalize. I'll dig a bit more on this and create another issue if I can't find a fix for it.
